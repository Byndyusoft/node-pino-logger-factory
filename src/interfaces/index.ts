export * from "./pino-http-logger-factory.interface";
export * from "./pino-http-logger-options-builder.interface";
export * from "./pino-logger-factory.interface";
export * from "./pino-logger-options-builder.interface";
