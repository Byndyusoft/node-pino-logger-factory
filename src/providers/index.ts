export * from "./pino-http-logger-factory";
export * from "./pino-http-logger-options-builder";
export * from "./pino-logger-factory";
export * from "./pino-logger-options-builder";
