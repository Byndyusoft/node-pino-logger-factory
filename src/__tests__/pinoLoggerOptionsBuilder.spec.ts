/*
 * Copyright 2021 Byndyusoft
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { pino } from "pino";
import { stdout } from "stdout-stderr";

import { LogLevel, PinoLoggerOptionsBuilder } from "~/src";

describe("PinoLoggerOptionsBuilder", () => {
  afterEach(() => {
    delete process.env.CONFIG_ENV;
    delete process.env.BUILD_COMMIT;
  });

  it("must build options with defaults", () => {
    const builder = new PinoLoggerOptionsBuilder();

    expect(builder.build()).toMatchSnapshot();
  });

  it("must build options without defaults", () => {
    const builder = new PinoLoggerOptionsBuilder(false);

    expect(builder.build()).toMatchSnapshot();
  });

  it("must log custom error", () => {
    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();
    pino(builder.build()).error({ err: "some error" });
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must log error", () => {
    const error = new Error("some error");
    error.stack = "Error: some error";

    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();
    pino(builder.build()).error(error);
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must log error with custom message", () => {
    const error = new Error("some error");
    error.stack = "Error: some error";

    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();
    pino(builder.build()).error(error, "custom message");
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must log error without trace", () => {
    const error = new Error("some error");
    delete error.stack;

    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();
    pino(builder.build()).error(error);
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must log error with extra undefined argument", () => {
    const error = new Error("some error");
    error.stack = "Error: some error";

    //only for nest logger in 9.3.0 and later
    //first param is always undefined if not use params in log.error()
    //if params in log.error() is set, then first extraArg mustn't be undefined
    const extraArguments = [undefined];

    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();

    pino(builder.build()).error(
      {
        err: error,
        context: "some context",
      },
      ...extraArguments,
    );
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must log message", () => {
    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();
    pino(builder.build()).info("hello");
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must log message with args", () => {
    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();
    pino(builder.build()).info("hello %s", "world");
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must log object", () => {
    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();
    pino(builder.build()).info({ a: 1 });
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must log object and message", () => {
    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();
    pino(builder.build()).info({ a: 1 }, "hello");
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must log object and message with args", () => {
    const builder = new PinoLoggerOptionsBuilder();

    stdout.start();
    pino(builder.build()).info({ a: 1 }, "hello %s", "world");
    stdout.stop();

    expect(JSON.parse(stdout.output)).toMatchSnapshot();
  });

  it("must override level", () => {
    const builder = new PinoLoggerOptionsBuilder();

    expect(builder.withLevel(LogLevel.warn).build()).toMatchSnapshot();
  });

  it("must override pretty print", () => {
    const builder = new PinoLoggerOptionsBuilder();

    expect(builder.withPrettyPrint(true).build()).toMatchSnapshot();
  });

  it("must use process.env that starts with BUILD_", () => {
    process.env.BUILD_COMMIT = "eefc20";

    const builder = new PinoLoggerOptionsBuilder();

    expect(builder.build()).toMatchSnapshot();
  });

  it("must use process.env.CONFIG_ENV if exists", () => {
    process.env.CONFIG_ENV = "dev";

    const builder = new PinoLoggerOptionsBuilder();

    expect(builder.build()).toMatchSnapshot();
  });
});
