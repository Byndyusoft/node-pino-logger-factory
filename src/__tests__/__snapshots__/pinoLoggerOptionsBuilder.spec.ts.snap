// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinoLoggerOptionsBuilder must build options with defaults 1`] = `
{
  "base": {
    "env": "test",
    "hostname": "host",
    "name": "name",
    "pid": 123456,
    "version": "version",
  },
  "formatters": {
    "level": [Function],
  },
  "hooks": {
    "logMethod": [Function],
  },
  "level": "info",
  "redact": [
    "req.headers.authorization",
    "err.config.headers['x-api-key']",
  ],
  "serializers": {
    "err": [Function],
    "req": [Function],
    "res": [Function],
  },
  "timestamp": [Function],
  "transport": undefined,
}
`;

exports[`PinoLoggerOptionsBuilder must build options without defaults 1`] = `
{
  "base": {},
  "formatters": {
    "level": [Function],
  },
  "hooks": {
    "logMethod": [Function],
  },
  "level": "info",
  "redact": [],
  "serializers": {},
  "timestamp": [Function],
  "transport": undefined,
}
`;

exports[`PinoLoggerOptionsBuilder must log custom error 1`] = `
{
  "env": "test",
  "err": "some error",
  "hostname": "host",
  "level": "error",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must log error 1`] = `
{
  "env": "test",
  "err": {
    "message": "some error",
    "stack": "Error: some error",
    "type": "Error",
  },
  "errHash": "932f64f7",
  "hostname": "host",
  "level": "error",
  "msg": "some error",
  "msgTemplateHash": "7611b3bb",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must log error with custom message 1`] = `
{
  "env": "test",
  "err": {
    "message": "some error",
    "stack": "Error: some error",
    "type": "Error",
  },
  "errHash": "932f64f7",
  "hostname": "host",
  "level": "error",
  "msg": "custom message",
  "msgTemplateHash": "90c7844f",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must log error with extra undefined argument 1`] = `
{
  "context": "some context",
  "env": "test",
  "err": {
    "message": "some error",
    "stack": "Error: some error",
    "type": "Error",
  },
  "errHash": "932f64f7",
  "hostname": "host",
  "level": "error",
  "msg": "some error",
  "msgTemplateHash": "7611b3bb",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must log error without trace 1`] = `
{
  "env": "test",
  "err": {
    "message": "some error",
    "stack": "",
    "type": "Error",
  },
  "errHash": "0",
  "hostname": "host",
  "level": "error",
  "msg": "some error",
  "msgTemplateHash": "7611b3bb",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must log message 1`] = `
{
  "env": "test",
  "hostname": "host",
  "level": "info",
  "msg": "hello",
  "msgTemplateHash": "248bfa47",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must log message with args 1`] = `
{
  "env": "test",
  "hostname": "host",
  "level": "info",
  "msg": "hello world",
  "msgTemplateHash": "3504eb44",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must log object 1`] = `
{
  "a": 1,
  "env": "test",
  "hostname": "host",
  "level": "info",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must log object and message 1`] = `
{
  "a": 1,
  "env": "test",
  "hostname": "host",
  "level": "info",
  "msg": "hello",
  "msgTemplateHash": "248bfa47",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must log object and message with args 1`] = `
{
  "a": 1,
  "env": "test",
  "hostname": "host",
  "level": "info",
  "msg": "hello world",
  "msgTemplateHash": "3504eb44",
  "name": "name",
  "pid": 123456,
  "time": "2016-04-05T17:02:19.796Z",
  "version": "version",
}
`;

exports[`PinoLoggerOptionsBuilder must override level 1`] = `
{
  "base": {
    "env": "test",
    "hostname": "host",
    "name": "name",
    "pid": 123456,
    "version": "version",
  },
  "formatters": {
    "level": [Function],
  },
  "hooks": {
    "logMethod": [Function],
  },
  "level": "warn",
  "redact": [
    "req.headers.authorization",
    "err.config.headers['x-api-key']",
  ],
  "serializers": {
    "err": [Function],
    "req": [Function],
    "res": [Function],
  },
  "timestamp": [Function],
  "transport": undefined,
}
`;

exports[`PinoLoggerOptionsBuilder must override pretty print 1`] = `
{
  "base": {
    "env": "test",
    "hostname": "host",
    "name": "name",
    "pid": 123456,
    "version": "version",
  },
  "formatters": {
    "level": [Function],
  },
  "hooks": {
    "logMethod": [Function],
  },
  "level": "info",
  "redact": [
    "req.headers.authorization",
    "err.config.headers['x-api-key']",
  ],
  "serializers": {
    "err": [Function],
    "req": [Function],
    "res": [Function],
  },
  "timestamp": [Function],
  "transport": {
    "options": {
      "colorize": true,
      "ignore": "pid,hostname,name,version,env,errHash,msgTemplateHash",
      "translateTime": "SYS:STANDARD",
    },
    "target": "pino-pretty",
  },
}
`;

exports[`PinoLoggerOptionsBuilder must use process.env that starts with BUILD_ 1`] = `
{
  "base": {
    "commit": "eefc20",
    "env": "test",
    "hostname": "host",
    "name": "name",
    "pid": 123456,
    "version": "version",
  },
  "formatters": {
    "level": [Function],
  },
  "hooks": {
    "logMethod": [Function],
  },
  "level": "info",
  "redact": [
    "req.headers.authorization",
    "err.config.headers['x-api-key']",
  ],
  "serializers": {
    "err": [Function],
    "req": [Function],
    "res": [Function],
  },
  "timestamp": [Function],
  "transport": undefined,
}
`;

exports[`PinoLoggerOptionsBuilder must use process.env.CONFIG_ENV if exists 1`] = `
{
  "base": {
    "env": "dev",
    "hostname": "host",
    "name": "name",
    "pid": 123456,
    "version": "version",
  },
  "formatters": {
    "level": [Function],
  },
  "hooks": {
    "logMethod": [Function],
  },
  "level": "info",
  "redact": [
    "req.headers.authorization",
    "err.config.headers['x-api-key']",
  ],
  "serializers": {
    "err": [Function],
    "req": [Function],
    "res": [Function],
  },
  "timestamp": [Function],
  "transport": undefined,
}
`;
